File: langs/csharp/src/VMx/Forms/FormVMBuilder.cs
using VMx.Builders;
using VMx.Internal;
using VMx.Services;

namespace VMx.Forms;

/// <summary>
/// Immutable fluent builder for <see cref="FormVM{TM}"/>.
/// Each setter returns a new builder instance (BLD-001).
/// <see cref="Build"/> validates required fields and raises
/// <see cref="BuilderValidationException"/> on failure (BLD-002).
/// See ADR-0035 §2 FV1 / FV2 and spec/10-builders.md §3.
/// Use <c>FormVM&lt;TM&gt;.Builder()</c> to start.
/// </summary>
/// <typeparam name="TM">Domain model type. Record types are recommended for structural equality.</typeparam>
public sealed class FormVMBuilder<TM>
    where TM : notnull
{
    // ── Required ──────────────────────────────────────────────────────────────
    private readonly TM? _initial;
    private readonly bool _initialSet;
    private readonly Func<TM, Task>? _persister;

    // ── Optional ──────────────────────────────────────────────────────────────
    private readonly IMessageHub? _hub;
    private readonly bool _strict;
    private readonly Func<TM, TM>? _snapshotter;
    private readonly Dictionary<string, Func<TM, string?>> _validators;
    private readonly Func<TM, IReadOnlyDictionary<string, string?>>? _modelValidator;
    private readonly Func<TM, TM>? _resetOnApproved;

    /// <summary>Empty starting builder.</summary>
    public static readonly FormVMBuilder<TM> Empty = new();

    private FormVMBuilder()
    {
        _validators = new Dictionary<string, Func<TM, string?>>();
    }

    private FormVMBuilder(
        TM? initial,
        bool initialSet,
        Func<TM, Task>? persister,
        IMessageHub? hub,
        bool strict,
        Func<TM, TM>? snapshotter,
        Dictionary<string, Func<TM, string?>> validators,
        Func<TM, IReadOnlyDictionary<string, string?>>? modelValidator,
        Func<TM, TM>? resetOnApproved)
    {
        _initial = initial;
        _initialSet = initialSet;
        _persister = persister;
        _hub = hub;
        _strict = strict;
        _snapshotter = snapshotter;
        _validators = validators;
        _modelValidator = modelValidator;
        _resetOnApproved = resetOnApproved;
    }

    // ── Fluent setters ───────────────────────────────────────────────────────

    /// <summary>Sets the required initial domain model.</summary>
    public FormVMBuilder<TM> Initial(TM initial) => With(initial: initial, initialSet: true);

    /// <summary>Sets the required async persister delegate <c>(model) -&gt; Task</c>.</summary>
    public FormVMBuilder<TM> Persister(Func<TM, Task> persister) => With(persister: persister);

    /// <summary>
    /// C#-only convenience: wires the typed-interface persister overload by
    /// delegating to the <see cref="Func{TM, Task}"/> form via
    /// <see cref="IFormPersister{TM}.PersistAsync"/>.
    /// </summary>
    public FormVMBuilder<TM> WithFormPersister(IFormPersister<TM> persister)
    {
        ThrowHelper.ThrowIfNull(persister, nameof(persister));
        return With(persister: model => persister.PersistAsync(model));
    }

    /// <summary>Sets the optional message hub (default: <see cref="NullMessageHub.Instance"/>).</summary>
    public FormVMBuilder<TM> Hub(IMessageHub hub) => With(hub: hub);

    /// <summary>
    /// Enables strict mode: <c>ApproveCommand.CanExecute()</c> returns
    /// <c>false</c> when <see cref="FormVM{TM}.IsDirty"/> is <c>false</c>.
    /// Default: <c>false</c>.
    /// </summary>
    public FormVMBuilder<TM> Strict(bool strict) => With(strict: strict);

    /// <summary>
    /// Sets a custom snapshot function (default: a deep clone via a
    /// <see cref="System.Text.Json"/> round-trip; inject this as the escape
    /// hatch for models JSON cannot round-trip).
    /// </summary>
    public FormVMBuilder<TM> Snapshotter(Func<TM, TM> snapshotter) => With(snapshotter: snapshotter);

    /// <summary>Registers a field validator. The returned builder is a new instance.</summary>
    public FormVMBuilder<TM> Validator(string field, Func<TM, string?> validator)
    {
        var validators = new Dictionary<string, Func<TM, string?>>(_validators)
        {
            [field] = validator,
        };
        return With(validators: validators);
    }

    /// <summary>Registers a model-level validator returning field-name errors.</summary>
    public FormVMBuilder<TM> ModelValidator(
        Func<TM, IReadOnlyDictionary<string, string?>> modelValidator)
        => With(modelValidator: modelValidator);

    /// <summary>
    /// Sets an optional post-persist callback that derives the next pristine
    /// model from the captured approved value.
    /// </summary>
    public FormVMBuilder<TM> ResetOnApproved(Func<TM, TM> resetOnApproved)
        => With(resetOnApproved: resetOnApproved);

    // ── Build ────────────────────────────────────────────────────────────────

    /// <summary>
    /// Validates required fields and constructs a <see cref="FormVM{TM}"/>.
    /// </summary>
    /// <exception cref="BuilderValidationException">
    /// Thrown when <c>Initial</c> or <c>Persister</c> has not been set.
    /// </exception>
    public FormVM<TM> Build()
    {
        if (!_initialSet) throw new BuilderValidationException("Initial");
        BuilderValidationException.Require(_persister, "Persister");

        return new FormVM<TM>(
            _initial!,
            _persister,
            _hub,
            _strict,
            _snapshotter,
            _validators,
            _modelValidator,
            _resetOnApproved);
    }

    // ── Wither ───────────────────────────────────────────────────────────────

    private FormVMBuilder<TM> With(
        TM? initial = default,
        bool? initialSet = null,
        Func<TM, Task>? persister = null,
        IMessageHub? hub = null,
        bool? strict = null,
        Func<TM, TM>? snapshotter = null,
        Dictionary<string, Func<TM, string?>>? validators = null,
        Func<TM, IReadOnlyDictionary<string, string?>>? modelValidator = null,
        Func<TM, TM>? resetOnApproved = null)
        => new(
            initialSet == true ? initial : _initial,
            initialSet ?? _initialSet,
            persister ?? _persister,
            hub ?? _hub,
            strict ?? _strict,
            snapshotter ?? _snapshotter,
            validators ?? _validators,
            modelValidator ?? _modelValidator,
            resetOnApproved ?? _resetOnApproved);
}

File: langs/swift/Sources/VMx/Builders/FormVMBuilder.swift
//
// FormVMBuilder<Model> — immutable fluent builder for FormVM<Model>.
//
// See spec/10-builders.md §3 and ADR-0035 §2 FV1/FV2.
//
// Required setters: initial, persister.
// Optional setters: hub, strict, snapshotter, equals.
//
// Each setter returns a NEW builder instance (BLD-001). build() validates
// required fields and throws BuilderValidationError naming the first missing
// one — "initial" before "persister" (BLD-002).
//

public struct FormVMBuilder<Model> {

    // ── Stored configuration ──────────────────────────────────────────────────

    private var _initial: InitialBox?
    private var _persister: ((Model) async throws -> Void)?
    private var _hub: MessageHubProtocol?
    private var _strict: Bool = false
    private var _snapshotter: ((Model) -> Model)?
    private var _equals: ((Model, Model) -> Bool)?

    /// One-shot box to distinguish "initial was assigned (possibly a nil-able
    /// generic)" from "initial was never set", without requiring Hashable.
    private struct InitialBox { let value: Model }

    // ── Init ──────────────────────────────────────────────────────────────────

    public init() {}

    // ── Setters (each returns a new builder copy) ─────────────────────────────

    /// Set the required initial domain model (also becomes the initial snapshot).
    public func initial(_ value: Model) -> FormVMBuilder<Model> {
        var copy = self; copy._initial = InitialBox(value: value); return copy
    }

    /// Set the required async persister `(Model) async throws -> Void`.
    public func persister(
        _ fn: @escaping (Model) async throws -> Void
    ) -> FormVMBuilder<Model> {
        var copy = self; copy._persister = fn; return copy
    }

    /// Set the optional message hub. Default: `NullMessageHub.INSTANCE`.
    public func hub(_ value: MessageHubProtocol) -> FormVMBuilder<Model> {
        var copy = self; copy._hub = value; return copy
    }

    /// Enable strict mode. Default: `false`.
    ///
    /// When `true`, `approveCommand.canExecute()` returns `false` while
    /// `isDirty == false`.
    public func strict(_ value: Bool) -> FormVMBuilder<Model> {
        var copy = self; copy._strict = value; return copy
    }

    /// Set a custom snapshot function.
    /// Default: identity `{ $0 }` (correct for Swift value types).
    public func snapshotter(
        _ fn: @escaping (Model) -> Model
    ) -> FormVMBuilder<Model> {
        var copy = self; copy._snapshotter = fn; return copy
    }

    /// Set a custom dirty-tracking equality predicate.
    /// Default (non-Equatable): `_defaultModelEquals` (reference identity for
    /// classes; "always changed" for non-Equatable value types). When `Model`
    /// conforms to `Equatable`, the constrained `build()` defaults to `==`.
    public func equals(
        _ fn: @escaping (Model, Model) -> Bool
    ) -> FormVMBuilder<Model> {
        var copy = self; copy._equals = fn; return copy
    }

    /// Pre-wire the hub to `NullMessageHub.INSTANCE`.
    /// Equivalent to `.hub(NullMessageHub.INSTANCE)`.
    public func withDefaultServices() -> FormVMBuilder<Model> {
        hub(NullMessageHub.INSTANCE)
    }

    // ── build() ───────────────────────────────────────────────────────────────

    /// Validate required fields and construct a `FormVM<Model>`.
    ///
    /// - Throws: `BuilderValidationError(missingField: "initial")` when the
    ///   initial model has not been set.
    /// - Throws: `BuilderValidationError(missingField: "persister")` when no
    ///   persister has been set.
    public func build() throws -> FormVM<Model> {
        try _buildCore()
    }

    // ── Internal ──────────────────────────────────────────────────────────────

    fileprivate func _buildCore() throws -> FormVM<Model> {
        guard let box = _initial else {
            throw BuilderValidationError(missingField: "initial")
        }
        guard let persister = _persister else {
            throw BuilderValidationError(missingField: "persister")
        }
        let hub = _hub ?? NullMessageHub.INSTANCE
        let snapshotter = _snapshotter ?? { $0 }
        let equals = _equals ?? { _defaultModelEquals($0, $1) }
        return FormVM(
            initial: box.value,
            persister: persister,
            hub: hub,
            strict: _strict,
            snapshotter: snapshotter,
            equals: equals
        )
    }
}

// ─── Equatable-constrained build() ───────────────────────────────────────────
//
// When Model: Equatable and no explicit equals predicate is set, default to ==
// (structural equality) — matching the FormVM Equatable convenience init.
// An explicit .equals(_:) still overrides (check _equals before defaulting).

extension FormVMBuilder where Model: Equatable {
    /// Equatable-aware `build()` — defaults `equals` to `==` so builder
    /// callers don't need to supply an explicit predicate for Equatable models.
    public func build() throws -> FormVM<Model> {
        if _equals == nil {
            return try equals(==)._buildCore()
        }
        return try _buildCore()
    }
}
